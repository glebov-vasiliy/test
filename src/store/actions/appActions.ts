import { createAction } from 'typesafe-actions'
import { Person, Tables } from '../types/appState'

export const addItem = createAction('@@app/addItem')<Person>()
export const saveItem = createAction('@@app/saveItem')<{ tableId: number; person: Person }>()
export const setEditableItem = createAction('@@app/setEditableItem')<
  { tableId: number; personId: number } | undefined
>()
export const updateTable = createAction('@@app/updateTable')<{ tableId: number; persons: Person[] }>()
export const updateTables = createAction('@@app/updateTables')<Tables>()
