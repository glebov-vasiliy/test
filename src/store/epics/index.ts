import { combineEpics } from 'redux-observable'
import appEpics from './appEpics'

export const rootEpics = combineEpics(...Object.values(appEpics))
