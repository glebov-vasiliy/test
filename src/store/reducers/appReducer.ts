import { AppState } from '../types/appState'
import * as Actions from '../actions'
import { createReducer } from 'typesafe-actions'
import { RootAction } from '../types'

const initialState: AppState = { defaultPerson: { id: 0, name: '', surname: '', age: '', city: '' }, tables: [[]] }
export const appReducer = createReducer<AppState, RootAction>(initialState)
  .handleAction(Actions.addItem, (state, { payload: person }) => ({
    ...state,
    tables: [
      ...state.tables.map((value, index) =>
        index ? value : [{ ...person, id: state.tables[0].length + 1 }, ...state.tables[0]],
      ),
    ],
  }))
  .handleAction(Actions.saveItem, (state, { payload: { person, tableId } }) => ({
    ...state,
    tables: [
      ...state.tables.map((value, index) =>
        index === tableId ? [...state.tables[0].map((p) => (p.id === person.id ? person : p))] : value,
      ),
    ],
  }))
  .handleAction(Actions.updateTable, (state, { payload: { tableId, persons } }) => ({
    ...state,
    tables: state.tables.map((table, index) => (index === tableId ? persons : table)),
  }))
  .handleAction(Actions.updateTables, (state, { payload: tables }) => ({
    ...state,
    tables,
  }))
  .handleAction(Actions.setEditableItem, (state, { payload: editableItem }) => ({
    ...state,
    editableItem,
  }))
