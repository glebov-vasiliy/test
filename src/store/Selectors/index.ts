import { RootState } from '../reducers'
import { Person, Tables } from '../types/appState'
import { createSelector } from 'reselect'

export const tablesSelector = ({ appState }: RootState): Tables => appState.tables
export const editableItemSelector = ({ appState }: RootState): { tableId: number; personId: number } | undefined =>
  appState.editableItem
export const defaultPersonSelector = ({ appState }: RootState): Person => appState.defaultPerson
export const editablePersonSelector = createSelector(
  editableItemSelector,
  tablesSelector,
  (item, tables) =>
    item && { person: tables[item.tableId].find((person) => person.id === item.personId), tableId: item.tableId },
)
