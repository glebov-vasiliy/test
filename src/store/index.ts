import { createStore, compose, applyMiddleware } from 'redux'
import { createEpicMiddleware } from 'redux-observable'
import { composeWithDevTools } from 'redux-devtools-extension'
import { rootReducer, RootState } from './reducers'
import { RootAction } from './types'
import { rootEpics } from './epics'

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose
  }
}
const epicMiddleware = createEpicMiddleware<RootAction, RootAction, RootState>({})
const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(epicMiddleware)))
epicMiddleware.run(rootEpics)
export default store
