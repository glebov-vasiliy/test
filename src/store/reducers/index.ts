import { combineReducers } from 'redux'
import { appReducer } from './appReducer'
import { AppState } from '../types/appState'

export const rootReducer = combineReducers({
  appState: appReducer,
})

export type RootState = { appState: AppState }
