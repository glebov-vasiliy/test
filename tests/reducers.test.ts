import { appReducer } from '../src/store/reducers/appReducer'
import { addItem, saveItem } from '../src/store/actions'
import * as Fixture from './__fixtures__/app'

describe('app Reducer', () => {
  test('Should be return the addItem state', () => {
    expect(appReducer(Fixture.defaultStateFixture, addItem(Fixture.addItemFixture))).toMatchSnapshot()
  })
  test('Should be return the saveItem state', () => {
    expect(appReducer(Fixture.defaultStateFixture, saveItem(Fixture.saveItemFixture))).toMatchSnapshot()
  })
})
