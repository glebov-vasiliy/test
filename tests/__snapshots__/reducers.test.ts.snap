// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app Reducer Should be return the addItem state 1`] = `
Object {
  "defaultPerson": Object {
    "age": "",
    "city": "",
    "id": 0,
    "name": "",
    "surname": "",
  },
  "tables": Array [
    Array [
      Object {
        "age": "22",
        "city": "city",
        "id": 2,
        "name": "name",
        "surname": "surname",
      },
      Object {
        "age": "22",
        "city": "city",
        "id": 1,
        "name": "name",
        "surname": "surname",
      },
    ],
  ],
}
`;

exports[`app Reducer Should be return the saveItem state 1`] = `
Object {
  "defaultPerson": Object {
    "age": "",
    "city": "",
    "id": 0,
    "name": "",
    "surname": "",
  },
  "tables": Array [
    Array [
      Object {
        "age": "23",
        "city": "city1",
        "id": 1,
        "name": "name1",
        "surname": "surname1",
      },
    ],
  ],
}
`;
